scarb/src/core/lockfile.rs

```diff
@@ -0,0 +1,251 @@
+pub struct PackageLock {
+    pub id: PackageId,
+    pub dependencies: BTreeSet<PackageName>,
+}
+
+mod serde_ext {
+    use crate::core::{PackageName, SourceId};
+    use semver::Version;
+    use serde::{Deserialize, Serialize};
+    use std::collections::BTreeSet;
+
+    #[derive(Debug, Eq, PartialEq, Ord, PartialOrd, Serialize, Deserialize)]
+    #[serde(rename_all = "snake_case")]
+    pub struct PackageLock {
+        pub name: PackageName,
+        pub version: Version,
+        #[serde(default = "SourceId::empty_path")]
+        #[serde(skip_serializing_if = "SourceId::is_path")]
+        pub source: SourceId,
+        #[serde(default = "BTreeSet::new")]
+        #[serde(skip_serializing_if = "BTreeSet::is_empty")]
+        pub dependencies: BTreeSet<PackageName>,
+    }
+}
+
+impl From<serde_ext::PackageLock> for PackageLock {
+    fn from(value: serde_ext::PackageLock) -> Self {
+        Self {
+            dependencies: value.dependencies,
+            id: PackageId::new(value.name, value.version, value.source),
+        }
+    }
+}
+
+impl From<PackageLock> for serde_ext::PackageLock {
+    fn from(value: PackageLock) -> Self {
+        Self {
+            dependencies: value.dependencies,
+            name: value.id.name.clone(),
+            version: value.id.version.clone(),
+            source: value.id.source_id,
+        }
+    }
+}
+
+impl Lockfile {
+    pub fn new(packages: impl Iterator<Item = PackageLock>) -> Self {
+        Self {
+            version: Default::default(),
+            packages: packages.collect(),
+        }
+    }
+
+    pub fn from_resolve(resolve: &Resolve) -> Self {
+        let packages = resolve.package_ids().map(|package| {
+            let deps = resolve
+                .package_dependencies(package)
+                .map(|dep| dep.name.clone());
+            PackageLock::new(&package, deps)
+        });
+        Self::new(packages)
+    }
+
+    fn header(&self) -> String {
+        "# Code generated by scarb DO NOT EDIT.".into()
+    }
+
+    fn body(&self) -> Result<Document> {
+        let toml = toml::Table::try_from(self)?;
+        let mut doc = toml.to_string().parse::<Document>()?;
+
+        for packs in doc["package"].as_array_of_tables_mut().iter_mut() {
+            for pack in packs.iter_mut() {
+                if let Some(deps) = pack.remove("dependencies") {
+                    let mut deps: Array = deps
+                        .into_value()
+                        .unwrap()
+                        .as_array()
+                        .unwrap()
+                        .iter()
+                        .map(|dep| {
+                            let mut dep = dep.clone();
+                            dep.decor_mut().set_prefix("\n ");
+                            dep
+                        })
+                        .collect();
+                    deps.set_trailing(",\n");
+                    pack.insert("dependencies", Item::Value(Value::Array(deps)));
```

Comment:
nitpick, aren't you able to mutate this array in-place using `*_mut` methods? https://docs.rs/toml_edit/latest/toml_edit/struct.Array.html?search=as_array_mut etc